### PR TITLE
Adding logic to delete load balancer

### PR DESCRIPTION
**Pull request by nikhiljindal**
Adding `mci delete` command to delete the load balancer created using `mci create`.

cc @csbell @G-Harmon @madhusudancs

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/13)

### CODE REVIEW

File: app/mci/cmd/delete.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+// Deletes the given ingress in the given list of clusters.
+func deleteIngressInClusters(kubeconfig, ingressFilename string, clusters []string) error {
+	kubectlArgs := []string{"kubectl"}
+	if kubeconfig != "" {
+		kubectlArgs = append(kubectlArgs, fmt.Sprintf("--kubeconfig=%s", kubeconfig))
+	}
+	// TODO(nikhiljindal): Validate and optionally add the gce-multi-cluster class annotation to the ingress YAML spec.
+	deleteArgs := append(kubectlArgs, []string{"delete", fmt.Sprintf("--filename=%s", ingressFilename)}...)
```
Comment (nikhiljindal) [inline app/mci/cmd/delete.go]: Will update this to use the go client instead of kubectl when https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/12 merges.

**Comment by G-Harmon** [comment]
Reviewed 27 of 27 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[app/mci/cmd/delete.go, line 35 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUIybxfv3nZMJ0ACSk:-KxUIybxfv3nZMJ0ACSl:b-peckbh) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L35)):*
> cluster
nit: should we s/clusters/contexts/? Since that's what we pass in. but then again it didn't really work trying to have multiple ingresses in the same cluster w/ different namespaces.

---

*[app/mci/cmd/delete.go, line 43 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJHo6Zmcyx-HdKPds:-KxUJHo6Zmcyx-HdKPdt:bk4opob) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L43)):*
> ```Go
> 	IngressFilename string
> 	// Path to kubeconfig.
> 	Kubeconfig string
> ```

it's inconsistent to have 'filename' appended to Ingress, but not to Kubeconfig. make this KubeconfigFilename?

---

*[app/mci/cmd/delete.go, line 78 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJb262UQYS6snrLVs:-KxUJb262UQYS6snrLVt:bnqrsxf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L78)):*
> ```Go
> 	cmd.Flags().StringVarP(&options.IngressFilename, "ingress", "i", options.IngressFilename, "filename containing ingress spec")
> 	cmd.Flags().StringVarP(&options.Kubeconfig, "kubeconfig", "k", options.Kubeconfig, "path to kubeconfig")
> 	cmd.Flags().StringVarP(&options.GCPProject, "gcp-project", "", options.GCPProject, "name of the gcp project")
> ```

How about adding a short flag for the project? like "-p"?

---

*[app/mci/cmd/delete.go, line 151 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKJomHgjsp6EAYzcn:-KxUKJomHgjsp6EAYzco:ba8y1nr) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L151)):*
> ```Go
> 			// TODO(nikhiljindal): Continue if this is an ingress does not exist error.
> 			glog.V(2).Infof("error in running command: %s", strings.Join(contextArgs, " "))
> 			return fmt.Errorf("error in deleting ingress in cluster %s: %s, output: %s", c, err, output)
> ```

Is it worth continuing regardless of what kind of error? In other words, what sort of error do you think we should stop for?

---

*[app/mci/cmd/delete_test.go, line 63 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKed3yHQNNw1cigxf:-KxUKed3yHQNNw1cigxg:b-9giwm5) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete_test.go#L63)):*
> ```Go
> }
> 
> func TestDeleteIngress(t *testing.T) {
> ```

worth adding a test where we DO hit an error deleting?
That might be hard now and easier when we're using the golang client. add a TODO for that test?

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxULZyU77SIeMBM0uxP:-KxULZyU77SIeMBM0uxQ:b-pa0wvd) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L90)):*
> ```Go
> 	glog.V(2).Infof("Deleting backend service %s", name)
> 	if err := b.bsp.DeleteGlobalBackendService(name); err != nil {
> 		glog.V(2).Infof("Error in deleting backend service %s: %s", name, err)
> ```

Infof or Warningf for errors?

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer_test.go, line 58 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUMLq6s2lTef1LdnZh:-KxUMLq6s2lTef1LdnZi:bvw012e) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer_test.go#L58)):*
> ```Go
> func TestDeleteForwardingRule(t *testing.T) {
> 	lbName := "lb-name"
> 	ipAddr := "192.168.0.0"
> ```

okay, one bad IP address is okay, but two is a pattern :)

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 144 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUMp1q9oXkYDKUbuaX:-KxUMp1q9oXkYDKUbuaY:b-vxcoz8) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L144)):*
> ```Go
> // DeleteLoadBalancer deletes the GCP resources associated with the L7 GCP load balancer for the given ingress.
> func (l *LoadBalancerSyncer) DeleteLoadBalancer(ing *v1beta1.Ingress) error {
> 	// TODO: Dont require the ingress yaml from users. Just the name should be enough. We can fetch ingress YAML from one of the clusters.
> ```

TODO(username/Issue)

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 150 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUN8-4CEbmb0Mgpc20:-KxUN8-4CEbmb0Mgpc21:b-32ayii) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L150)):*
> ```Go
> 	// For ex: we create health checks before creating backend services (so that backend service can point to the health check),
> 	// but delete backend service before deleting health check. We cannot delete the health check when backend service is still pointing to it.
> 	// Create health checks to be used by the backend service.
> ```

stray comment? (Create health checks to be used by the backend service.)

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 242 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUNZT_3aWhTmyIq-Cg:-KxUNZT_3aWhTmyIq-Ch:b15djen) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L242)):*
> ```Go
> 	// Reserve a global address. User is supposed to do this before calling CreateLoadBalancer.
> 	lbc.ipp.ReserveGlobalAddress(ipAddress)
> 	// Create ingress with instance groups annotation.
> ```

Can you factor out the code to create a working Ingress? It's pretty long.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 342 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUNq0yrEQTmuMGRn0a:-KxUNq0yrEQTmuMGRn0b:b8k2mbc) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L342)):*
> ```Go
> 	}
> 	if len(ffr.EnsuredForwardingRules) != 0 {
> 		t.Errorf("unexpected forwarding rules have not been deleted")
> ```

Errorf should include the actual length or contents

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13)*

**Comment by nikhiljindal** [comment]
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[app/mci/cmd/delete.go, line 35 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUIybxfv3nZMJ0ACSk:-KxUg3fX1idd0PU72oZu:b-smyvzu) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L35)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> cluster
nit: should we s/clusters/contexts/? Since that's what we pass in. but then again it didn't really work trying to have multiple ingresses in the same cluster w/ different namespaces.
</blockquote></details>

Yes I prefer keeping clusters.
Cluster registry will also call it list of clusters.

---

*[app/mci/cmd/delete.go, line 43 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJHo6Zmcyx-HdKPds:-KxUfxQ6oPOrpyRQsDMa:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L43)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

it's inconsistent to have 'filename' appended to Ingress, but not to Kubeconfig. make this KubeconfigFilename?
</blockquote></details>

Done.

---

*[app/mci/cmd/delete.go, line 78 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJb262UQYS6snrLVs:-KxUgKRFI9FmLtDkCZw4:b-ps9fpy) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L78)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

How about adding a short flag for the project? like "-p"?
</blockquote></details>

I am hoping that we add logic to auto detect it so users wont have to specify it.
We can add a short flag later if required. 

---

*[app/mci/cmd/delete.go, line 151 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKJomHgjsp6EAYzcn:-KxUgc7bmfzydIutsbho:b-96fbmy) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L151)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Is it worth continuing regardless of what kind of error? In other words, what sort of error do you think we should stop for?
</blockquote></details>

None. Note that the code here does not stop in case of error. It appends all errors and returns them.
The TODO is to not even return an error if the ingress does not exist.

---

*[app/mci/cmd/delete_test.go, line 63 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKed3yHQNNw1cigxf:-KxUh195FV7pmjXcg8Fy:bj9c98c) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete_test.go#L63)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

worth adding a test where we DO hit an error deleting?
That might be hard now and easier when we're using the golang client. add a TODO for that test?
</blockquote></details>

Done. Yes generally we need better edge case and error testing.
Lot of our tests right now are for basic happy path.

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxULZyU77SIeMBM0uxP:-KxUhCXybsfArqMQ_BTk:bit17yp) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L90)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Infof or Warningf for errors?
</blockquote></details>

V(2) does not support Errorf

---

*[app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer_test.go, line 58 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUMLq6s2lTef1LdnZh:-KxUiZyuxTXEECOChPlO:b-6k4730) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/forwardingrule/forwardingrulesyncer_test.go#L58)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

okay, one bad IP address is okay, but two is a pattern :)
</blockquote></details>

haha fixed :)

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 144 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUMp1q9oXkYDKUbuaX:-KxUigzvo_woqSCOfvg7:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L144)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

TODO(username/Issue)
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 150 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUN8-4CEbmb0Mgpc20:-KxUimUMYOe1Q_PaKLET:b-hfpv0l) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L150)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

stray comment? (Create health checks to be used by the backend service.)
</blockquote></details>

removed

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 242 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUNZT_3aWhTmyIq-Cg:-KxUnQtWZ2GkjAbeintD:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L242)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Can you factor out the code to create a working Ingress? It's pretty long.
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go, line 342 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUNq0yrEQTmuMGRn0a:-KxUjYNp0c7I1bOHgXPo:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer_test.go#L342)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Errorf should include the actual length or contents
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13)*

**Comment by nikhiljindal** [comment]
Updated code as per comments. PTAL

**Comment by G-Harmon** [comment]
LGTM after figuring out one last thing about returning error vs grouping them while deleting.

---

Reviewed 5 of 6 files at r2.
Review status: 26 of 27 files reviewed at latest revision, 11 unresolved discussions.

---

*[app/mci/cmd/delete.go, line 78 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJb262UQYS6snrLVs:-KxV72EOnkpuxIVLHtLI:b-m3fbyb) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L78)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

I am hoping that we add logic to auto detect it so users wont have to specify it.
We can add a short flag later if required. 
</blockquote></details>

ah, got it. (should we add a comment here about that future direction?)

---

*[app/mci/cmd/delete.go, line 151 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKJomHgjsp6EAYzcn:-KxV7Q5q65sG08qgSm7Y:bdik1mi) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L151)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

None. Note that the code here does not stop in case of error. It appends all errors and returns them.
The TODO is to not even return an error if the ingress does not exist.
</blockquote></details>

hmm? The code says "return fmt.Errorf...". Changing it to append errors sgtm...

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxULZyU77SIeMBM0uxP:-KxV7_kKWNoKja8Y-_8r:bletb2t) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L90)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary><blockquote>

V(2) does not support Errorf
</blockquote></details>

I meant drop the V(2) and make it glog.Warningf. up to you though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-:-KxV8MPAx25TnZg8d3kb:b8tel42)*

**Comment by nikhiljindal** [comment]
Review status: 26 of 27 files reviewed at latest revision, 11 unresolved discussions.

---

*[app/mci/cmd/delete.go, line 78 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUJb262UQYS6snrLVs:-KxV9Ou5X8DZF--wkZVi:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L78)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

ah, got it. (should we add a comment here about that future direction?)
</blockquote></details>

Done.

---

*[app/mci/cmd/delete.go, line 151 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxUKJomHgjsp6EAYzcn:-KxV9nAo2KqwuLeT6tTr:bvsxgmg) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/cmd/delete.go#L151)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

hmm? The code says "return fmt.Errorf...". Changing it to append errors sgtm...
</blockquote></details>

aah got it. Fixed.

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 90 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-KxULZyU77SIeMBM0uxP:-KxV9v7OrP3bmIf1vRBP:b-xaj4k8) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/088a914215c161740a4426ac7d46d4ffaa86f8af/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L90)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

I meant drop the V(2) and make it glog.Warningf. up to you though.
</blockquote></details>

The parent function prints all the error messages.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13)*

**Comment by G-Harmon** [comment]
:lgtm:

---

Review status: 25 of 27 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/13#-:-KxVCaczfI1y_qN_x3SX:bnfp4nl)*

**Comment by nikhiljindal** [comment]
Thanks! Rebased to include https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/12
